File: python/packages/botas/src/botas/conversation_client.py
"""High-level client for the Bot Service Conversation REST API.

Wraps :class:`_BotHttpClient` to provide typed methods for sending,
updating, and deleting activities, managing conversation members, and
creating conversations.
"""

from __future__ import annotations

from typing import Any, Optional, Union
from urllib.parse import quote

from botas.bot_http_client import TokenProvider, _BotHttpClient, _BotRequestOptions
from botas.core_activity import (
    ChannelAccount,
    Conversation,
    CoreActivity,
    ResourceResponse,
    _ConversationParameters,
    _ConversationResourceResponse,
    _ConversationsResult,
    _PagedMembersResult,
    _Transcript,
)
from botas.meter_provider import get_metrics
from botas.tracer_provider import get_tracer


def _encode_conversation_id(conversation_id: str) -> str:
    return quote(conversation_id, safe="")


def _encode_id(id_value: str) -> str:
    """URL-encode an ID for use in path parameters."""
    return quote(id_value, safe="")


def _serialize(obj: Any) -> Any:
    """Convert Pydantic model or plain dict to a JSON-serializable dict."""
    if hasattr(obj, "model_dump"):
        return obj.model_dump(by_alias=True, exclude_none=True)
    return obj


class ConversationClient:
    """Typed client for Bot Service Conversation REST API operations.

    All methods accept a ``service_url`` and ``conversation_id`` to target
    the correct channel endpoint.  Authentication is handled automatically
    via the injected :class:`TokenProvider`.
    """

    def __init__(self, get_token: Optional[TokenProvider] = None) -> None:
        """Initialise the conversation client.

        Args:
            get_token: Async callable that supplies a bearer token.
                When ``None``, requests are unauthenticated.
        """
        self._http = _BotHttpClient(get_token)

    async def send_activity_async(
        self,
        service_url: str,
        conversation_id: str,
        activity: Union[
            CoreActivity,
            dict[str, Any],
        ],
    ) -> Optional[ResourceResponse]:
        """Send an activity to a conversation.

        Args:
            service_url: The channel's service URL.
            conversation_id: Target conversation identifier.
            activity: Activity payload (model or dict).

        Returns:
            A :class:`ResourceResponse` with the new activity ID, or ``None``.
        """
        tracer = get_tracer()
        metrics = get_metrics()
        if metrics:
            metrics.outbound_calls.add(1, {"operation": "sendActivity"})
        if tracer:
            with tracer.start_as_current_span("botas.conversation_client") as span:
                span.set_attribute("conversation.id", conversation_id)
                span.set_attribute("activity.type", getattr(activity, "type", "") or "")
                span.set_attribute("service.url", service_url)
                try:
                    result = await self._do_send(service_url, conversation_id, activity)
                    if result:
                        span.set_attribute("activity.id", result.id or "")
                    return result
                except Exception:
                    if metrics:
                        metrics.outbound_errors.add(1, {"operation": "sendActivity"})
                    raise
        try:
            return await self._do_send(service_url, conversation_id, activity)
        except Exception:
            if metrics:
                metrics.outbound_errors.add(1, {"operation": "sendActivity"})
            raise

    async def _do_send(
        self,
        service_url: str,
        conversation_id: str,
        activity: Union[
            CoreActivity,
            dict[str, Any],
        ],
    ) -> Optional[ResourceResponse]:
        """Execute the actual HTTP POST for sending an activity."""
        endpoint = f"/v3/conversations/{_encode_conversation_id(conversation_id)}/activities"
        data = await self._http.post(
            service_url,
            endpoint,
            _serialize(activity),
            _BotRequestOptions(operation_description="send activity"),
        )
        return ResourceResponse.model_validate(data) if data else None

    async def update_activity_async(
        self,
        service_url: str,
        conversation_id: str,
        activity_id: str,
        activity: Union[
            CoreActivity,
            dict[str, Any],
        ],
    ) -> Optional[ResourceResponse]:
        """Update an existing activity in a conversation.

        Args:
            service_url: The channel's service URL.
            conversation_id: Conversation containing the activity.
            activity_id: ID of the activity to update.
            activity: Replacement activity payload.

        Returns:
            A :class:`ResourceResponse`, or ``None``.
        """
        metrics = get_metrics()
        if metrics:
            metrics.outbound_calls.add(1, {"operation": "updateActivity"})
        try:
            endpoint = (
                f"/v3/conversations/{_encode_conversation_id(conversation_id)}/activities/{_encode_id(activity_id)}"
            )
            data = await self._http.put(
                service_url,
                endpoint,
                _serialize(activity),
                _BotRequestOptions(operation_description="update activity"),
            )
            return ResourceResponse.model_validate(data) if data else None
        except Exception:
            if metrics:
                metrics.outbound_errors.add(1, {"operation": "updateActivity"})
            raise

    async def delete_activity_async(self, service_url: str, conversation_id: str, activity_id: str) -> None:
        """Delete an activity from a conversation.

        Args:
            service_url: The channel's service URL.
            conversation_id: Conversation containing the activity.
            activity_id: ID of the activity to delete.
        """
        metrics = get_metrics()
        if metrics:
            metrics.outbound_calls.add(1, {"operation": "deleteActivity"})
        try:
            endpoint = (
                f"/v3/conversations/{_encode_conversation_id(conversation_id)}/activities/{_encode_id(activity_id)}"
            )
            await self._http.delete(
                service_url,
                endpoint,
                _BotRequestOptions(operation_description="delete activity"),
            )
        except Exception:
            if metrics:
                metrics.outbound_errors.add(1, {"operation": "deleteActivity"})
            raise

    async def get_conversation_members_async(self, service_url: str, conversation_id: str) -> list[ChannelAccount]:
        """Retrieve all members of a conversation.

        Args:
            service_url: The channel's service URL.
            conversation_id: Target conversation identifier.

        Returns:
            List of :class:`ChannelAccount` objects for each member.
        """
        endpoint = f"/v3/conversations/{_encode_conversation_id(conversation_id)}/members"
        data = await self._http.get(
            service_url,
            endpoint,
            options=_BotRequestOptions(operation_description="get conversation members"),
        )
        return [ChannelAccount.model_validate(m) for m in (data or [])]

    async def get_conversation_member_async(
        self, service_url: str, conversation_id: str, member_id: str
    ) -> Optional[ChannelAccount]:
        """Retrieve a single conversation member by ID.

        Args:
            service_url: The channel's service URL.
            conversation_id: Target conversation identifier.
            member_id: The member's account ID.

        Returns:
            A :class:`ChannelAccount`, or ``None`` if the member is not found.
        """
        endpoint = f"/v3/conversations/{_encode_conversation_id(conversation_id)}/members/{_encode_id(member_id)}"
        data = await self._http.get(
            service_url,
            endpoint,
            options=_BotRequestOptions(operation_description="get conversation member", return_none_on_not_found=True),
        )
        return ChannelAccount.model_validate(data) if data else None

    async def get_conversation_paged_members_async(
        self,
        service_url: str,
        conversation_id: str,
        page_size: Optional[int] = None,
        continuation_token: Optional[str] = None,
    ) -> _PagedMembersResult:
        """Retrieve conversation members with server-side pagination.

        Args:
            service_url: The channel's service URL.
            conversation_id: Target conversation identifier.
            page_size: Maximum members per page (channel may enforce its own limit).
            continuation_token: Opaque token from a previous page to fetch the next.

        Returns:
            A :class:`_PagedMembersResult` containing members and an optional
            continuation token for the next page.
        """
        endpoint = f"/v3/conversations/{_encode_conversation_id(conversation_id)}/pagedmembers"
        params = {
            "pageSize": str(page_size) if page_size else None,
            "continuationToken": continuation_token,
        }
        data = await self._http.get(
            service_url,
            endpoint,
            params=params,
            options=_BotRequestOptions(operation_description="get paged members"),
        )
        return _PagedMembersResult.model_validate(data) if data else _PagedMembersResult()

    async def delete_conversation_member_async(self, service_url: str, conversation_id: str, member_id: str) -> None:
        """Remove a member from a conversation.

        Args:
            service_url: The channel's service URL.
            conversation_id: Target conversation identifier.
            member_id: The member's account ID.
        """
        endpoint = f"/v3/conversations/{_encode_conversation_id(conversation_id)}/members/{_encode_id(member_id)}"
        await self._http.delete(
            service_url,
            endpoint,
            _BotRequestOptions(operation_description="delete conversation member"),
        )

    async def create_conversation_async(
        self, service_url: str, parameters: _ConversationParameters
    ) -> Optional[_ConversationResourceResponse]:
        """Create a new conversation on the channel.

        Args:
            service_url: The channel's service URL.
            parameters: Conversation creation parameters (members, topic, etc.).

        Returns:
            A :class:`_ConversationResourceResponse` with the new conversation
            ID and service URL, or ``None``.
        """
        data = await self._http.post(
            service_url,
            "/v3/conversations",
            _serialize(parameters),
            _BotRequestOptions(operation_description="create conversation"),
        )
        return _ConversationResourceResponse.model_validate(data) if data else None

    async def get_conversations_async(
        self, service_url: str, continuation_token: Optional[str] = None
    ) -> _ConversationsResult:
        """List conversations the bot has participated in.

        Args:
            service_url: The channel's service URL.
            continuation_token: Opaque token from a previous page.

        Returns:
            A :class:`_ConversationsResult` with conversations and an optional
            continuation token.
        """
        params = {"continuationToken": continuation_token}
        data = await self._http.get(
            service_url,
            "/v3/conversations",
            params=params,
            options=_BotRequestOptions(operation_description="get conversations"),
        )
        return _ConversationsResult.model_validate(data) if data else _ConversationsResult()

    async def send_conversation_history_async(
        self, service_url: str, conversation_id: str, _Transcript: _Transcript
    ) -> Optional[ResourceResponse]:
        """Upload a _Transcript of activities to a conversation's history.

        Args:
            service_url: The channel's service URL.
            conversation_id: Target conversation identifier.
            _Transcript: A :class:`_Transcript` containing activities to upload.

        Returns:
            A :class:`ResourceResponse`, or ``None``.
        """
        endpoint = f"/v3/conversations/{_encode_conversation_id(conversation_id)}/activities/history"
        data = await self._http.post(
            service_url,
            endpoint,
            _serialize(_Transcript),
            _BotRequestOptions(operation_description="send conversation history"),
        )
        return ResourceResponse.model_validate(data) if data else None

    async def get_conversation_account_async(self, service_url: str, conversation_id: str) -> Optional[Conversation]:
        """Retrieve the conversation account details.

        Args:
            service_url: The channel's service URL.
            conversation_id: Target conversation identifier.

        Returns:
            A :class:`Conversation` object, or ``None`` if not found.
        """
        endpoint = f"/v3/conversations/{_encode_conversation_id(conversation_id)}"
        data = await self._http.get(
            service_url,
            endpoint,
            options=_BotRequestOptions(operation_description="get conversation", return_none_on_not_found=True),
        )
        return Conversation.model_validate(data) if data else None

    async def aclose(self) -> None:
        """Close the underlying HTTP client and release resources."""
        await self._http.aclose()

File: node/packages/botas-core/src/conversation-client.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { _BotHttpClient, type TokenProvider } from './bot-http-client.js'
import { getLogger } from './logger.js'
import { getTracer } from './tracer-provider.js'
import { getMetrics } from './meter-provider.js'
import type {
  CoreActivity,
  ChannelAccount,
  Conversation,
  ConversationParameters,
  ConversationResourceResponse,
  ConversationsResult,
  PagedMembersResult,
  ResourceResponse,
  Transcript,
} from './core-activity.js'

/**
 * Client for the Bot Service v3 Conversations REST API.
 *
 * Sends activities, manages members, creates conversations, and uploads attachments.
 */
export class ConversationClient {
  private readonly http: _BotHttpClient

  /**
   * Create a new ConversationClient.
   *
   * @param getToken - Optional token provider for authenticating outbound API calls.
   *   Omit for unauthenticated local development.
   */
  constructor (getToken?: TokenProvider) {
    this.http = new _BotHttpClient(getToken)
  }

  /**
   * Send an activity to a conversation.
   *
   * @param serviceUrl - Bot Service service URL (from the incoming activity).
   * @param conversationId - Target conversation ID.
   * @param activity - CoreActivity payload to send.
   * @returns The {@link ResourceResponse} containing the new activity ID, or `undefined`.
   */
  async sendCoreActivityAsync (
    serviceUrl: string,
    conversationId: string,
    activity: Partial<CoreActivity>
  ): Promise<ResourceResponse | undefined> {
    const tracer = getTracer()
    const ccSpan = tracer?.startSpan('botas.conversation_client')
    ccSpan?.setAttribute('conversation.id', conversationId)
    ccSpan?.setAttribute('activity.type', (activity as CoreActivity).type ?? '')
    ccSpan?.setAttribute('service.url', serviceUrl)

    const metrics = getMetrics()
    metrics?.outboundApiCalls.add(1, { operation: 'sendActivity' })

    try {
      const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}/activities`
      getLogger().trace('Sending activity to %s%s', serviceUrl, endpoint)
      const result = await this.http.post<ResourceResponse>(
        serviceUrl,
        endpoint,
        activity,
        { operationDescription: 'send activity' }
      )
      ccSpan?.setAttribute('activity.id', result?.id ?? '')
      return result
    } catch (err) {
      ccSpan?.recordException(err instanceof Error ? err : new Error(String(err)))
      metrics?.outboundApiErrors.add(1, { operation: 'sendActivity' })
      throw err
    } finally {
      ccSpan?.end()
    }
  }

  /**
   * Update an existing activity in a conversation.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @param activityId - ID of the activity to update.
   * @param activity - Updated activity payload (partial merge).
   * @returns The {@link ResourceResponse}, or `undefined`.
   */
  async updateCoreActivityAsync (
    serviceUrl: string,
    conversationId: string,
    activityId: string,
    activity: Partial<CoreActivity>
  ): Promise<ResourceResponse | undefined> {
    const metrics = getMetrics()
    metrics?.outboundApiCalls.add(1, { operation: 'updateActivity' })
    const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}/activities/${activityId}`
    getLogger().trace('Updating activity at %s%s', serviceUrl, endpoint)
    try {
      return await this.http.put<ResourceResponse>(
        serviceUrl,
        endpoint,
        activity,
        { operationDescription: 'update activity' }
      )
    } catch (err) {
      metrics?.outboundApiErrors.add(1, { operation: 'updateActivity' })
      throw err
    }
  }

  /**
   * Delete an activity from a conversation.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @param activityId - ID of the activity to delete.
   */
  async deleteCoreActivityAsync (
    serviceUrl: string,
    conversationId: string,
    activityId: string
  ): Promise<void> {
    const metrics = getMetrics()
    metrics?.outboundApiCalls.add(1, { operation: 'deleteActivity' })
    const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}/activities/${activityId}`
    getLogger().trace('Deleting activity at %s%s', serviceUrl, endpoint)
    try {
      await this.http.delete(
        serviceUrl,
        endpoint,
        undefined,
        { operationDescription: 'delete activity' }
      )
    } catch (err) {
      metrics?.outboundApiErrors.add(1, { operation: 'deleteActivity' })
      throw err
    }
  }

  /**
   * Retrieve all members of a conversation.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @returns Array of {@link ChannelAccount} members; empty array if none.
   */
  async getConversationMembersAsync (
    serviceUrl: string,
    conversationId: string
  ): Promise<ChannelAccount[]> {
    const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}/members`
    getLogger().trace('Getting conversation members from %s%s', serviceUrl, endpoint)
    const result = await this.http.get<ChannelAccount[]>(
      serviceUrl,
      endpoint,
      undefined,
      { operationDescription: 'get conversation members' }
    )
    return result ?? []
  }

  /**
   * Retrieve a single member of a conversation by user ID.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @param memberId - ID of the member to retrieve.
   * @returns The {@link ChannelAccount}, or `undefined` if not found.
   */
  async getConversationMemberAsync (
    serviceUrl: string,
    conversationId: string,
    memberId: string
  ): Promise<ChannelAccount | undefined> {
    const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}/members/${memberId}`
    getLogger().trace('Getting conversation member from %s%s', serviceUrl, endpoint)
    return this.http.get<ChannelAccount>(
      serviceUrl,
      endpoint,
      undefined,
      { operationDescription: 'get conversation member', returnNullOnNotFound: true }
    )
  }

  /**
   * Retrieve a page of conversation members.
   *
   * Use `continuationToken` from the previous result to fetch subsequent pages.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @param pageSize - Maximum number of members per page.
   * @param continuationToken - Opaque token from a previous page result.
   * @returns A {@link PagedMembersResult} with members and an optional continuation token.
   */
  async getConversationPagedMembersAsync (
    serviceUrl: string,
    conversationId: string,
    pageSize?: number,
    continuationToken?: string
  ): Promise<PagedMembersResult<ChannelAccount>> {
    const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}/pagedmembers`
    getLogger().trace('Getting paged conversation members from %s%s', serviceUrl, endpoint)
    const params: Record<string, string | undefined> = {
      pageSize: pageSize?.toString(),
      continuationToken,
    }
    const result = await this.http.get<PagedMembersResult<ChannelAccount>>(
      serviceUrl,
      endpoint,
      params,
      { operationDescription: 'get paged members' }
    )
    return result ?? { members: [] }
  }

  /**
   * Remove a member from a conversation.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @param memberId - ID of the member to remove.
   */
  async deleteConversationMemberAsync (
    serviceUrl: string,
    conversationId: string,
    memberId: string
  ): Promise<void> {
    const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}/members/${memberId}`
    getLogger().trace('Deleting conversation member at %s%s', serviceUrl, endpoint)
    await this.http.delete(
      serviceUrl,
      endpoint,
      undefined,
      { operationDescription: 'delete conversation member' }
    )
  }

  /**
   * Create a new proactive conversation.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param parameters - Conversation creation parameters (members, topic, initial activity).
   * @returns A {@link ConversationResourceResponse} with the new conversation ID, or `undefined`.
   */
  async createConversationAsync (
    serviceUrl: string,
    parameters: ConversationParameters
  ): Promise<ConversationResourceResponse | undefined> {
    getLogger().trace('Creating conversation at %s', serviceUrl)
    return this.http.post<ConversationResourceResponse>(
      serviceUrl,
      '/v3/conversations',
      parameters,
      { operationDescription: 'create conversation' }
    )
  }

  /**
   * List all conversations the bot is a member of.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param continuationToken - Opaque token from a previous page result.
   * @returns A {@link ConversationsResult} with conversations and an optional continuation token.
   */
  async getConversationsAsync (
    serviceUrl: string,
    continuationToken?: string
  ): Promise<ConversationsResult> {
    getLogger().trace('Getting conversations from %s', serviceUrl)
    const result = await this.http.get<ConversationsResult>(
      serviceUrl,
      '/v3/conversations',
      { continuationToken },
      { operationDescription: 'get conversations' }
    )
    return result ?? { conversations: [] }
  }

  /**
   * Upload a transcript of past activities to a conversation.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @param transcript - Ordered list of activities to upload.
   * @returns The {@link ResourceResponse}, or `undefined`.
   */
  async sendConversationHistoryAsync (
    serviceUrl: string,
    conversationId: string,
    transcript: Transcript
  ): Promise<ResourceResponse | undefined> {
    const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}/activities/history`
    getLogger().trace('Sending conversation history to %s%s', serviceUrl, endpoint)
    return this.http.post<ResourceResponse>(
      serviceUrl,
      endpoint,
      transcript,
      { operationDescription: 'send conversation history' }
    )
  }

  /**
   * Retrieve the conversation account details.
   *
   * @param serviceUrl - Bot Service service URL.
   * @param conversationId - Target conversation ID.
   * @returns The {@link Conversation}, or `undefined` if not found.
   */
  async getConversationAsync (
    serviceUrl: string,
    conversationId: string
  ): Promise<Conversation | undefined> {
    const endpoint = `/v3/conversations/${encodeConversationId(conversationId)}`
    getLogger().trace('Getting conversation account from %s%s', serviceUrl, endpoint)
    return this.http.get<Conversation>(
      serviceUrl,
      endpoint,
      undefined,
      { operationDescription: 'get conversation', returnNullOnNotFound: true }
    )
  }
}

function encodeConversationId (conversationId: string): string {
  return encodeURIComponent(conversationId)
}
